File: server/api/routes/theme.py
"""Theme helpers — focal / daily background metadata + image byte proxy."""

from __future__ import annotations

from fastapi import APIRouter, Query
from fastapi.responses import Response

from server.api.deps import API_DEPS
from server.api.schemas.responses.theme import FocalBackgroundResponse
from server.services.focal_background import (
    fetch_focal_background,
    fetch_focal_background_image,
)

router = APIRouter(
    prefix="/api/v1/theme",
    tags=["theme"],
    dependencies=API_DEPS,
)


@router.get("/focal-background", response_model=FocalBackgroundResponse)
async def focal_background(
    locale: str | None = Query(
        default=None,
        max_length=32,
        description="UI locale (en / zh-Hans / zh-Hant) or Bing mkt (en-US).",
    ),
    idx: int = Query(
        default=0,
        ge=0,
        le=7,
        description="Bing HPImageArchive idx: 0=today, 1=yesterday, … up to 7.",
    ),
) -> FocalBackgroundResponse:
    """Proxy Bing HPImageArchive JSON; returns absolute https image URL + credits."""
    return await fetch_focal_background(locale, idx=idx)


@router.get("/focal-background/image")
async def focal_background_image(
    locale: str | None = Query(
        default=None,
        max_length=32,
        description="UI locale (en / zh-Hans / zh-Hant) or Bing mkt (en-US).",
    ),
    idx: int = Query(
        default=0,
        ge=0,
        le=7,
        description="Bing HPImageArchive idx: 0=today, 1=yesterday, … up to 7.",
    ),
) -> Response:
    """Proxy Bing wallpaper bytes for same-origin CSS / preview use."""
    body, media_type = await fetch_focal_background_image(locale, idx=idx)
    return Response(
        content=body,
        media_type=media_type,
        headers={"Cache-Control": "private, max-age=3600"},
    )

File: web/src/api/theme.ts
/**
 * Theme helpers — focal / daily background via server Bing proxy.
 */

import { apiClient } from "./client";
import type { components } from "./generated/schema";

export type FocalBackground = components["schemas"]["FocalBackgroundResponse"];

export type FocalBackgroundQuery = {
  locale?: string;
  /** Bing HPImageArchive idx: 0=today … 7 (recent days). */
  idx?: number;
};

function focalQueryParams(query?: FocalBackgroundQuery): Record<string, string> | undefined {
  if (!query) return undefined;
  const params: Record<string, string> = {};
  if (query.locale) params.locale = query.locale;
  if (query.idx != null && Number.isFinite(query.idx)) {
    params.idx = String(Math.max(0, Math.min(7, Math.trunc(query.idx))));
  }
  return Object.keys(params).length > 0 ? params : undefined;
}

/** Server proxies Bing HPImageArchive JSON (avoids browser CORS). */
export function fetchFocalBackground(
  localeOrQuery?: string | FocalBackgroundQuery,
): Promise<FocalBackground> {
  const query: FocalBackgroundQuery | undefined =
    typeof localeOrQuery === "string" ? { locale: localeOrQuery } : localeOrQuery;
  return apiClient.get<FocalBackground>(
    "/api/v1/theme/focal-background",
    focalQueryParams(query),
  );
}

/** Server proxies Bing wallpaper bytes (auth header; for blob CSS apply). */
export function fetchFocalBackgroundImage(
  localeOrQuery?: string | FocalBackgroundQuery,
): Promise<Blob> {
  const query: FocalBackgroundQuery | undefined =
    typeof localeOrQuery === "string" ? { locale: localeOrQuery } : localeOrQuery;
  const params = focalQueryParams(query);
  // Keep a static `/image` path prefix (optional `?...` after `?`) so route-inventory
  // scanners do not treat `${…}` as a path parameter glued onto `image`.
  const path = params
    ? `/api/v1/theme/focal-background/image?${new URLSearchParams(params).toString()}`
    : "/api/v1/theme/focal-background/image";
  return apiClient.getBlob(path, {
    timeoutMs: 20_000,
  });
}
